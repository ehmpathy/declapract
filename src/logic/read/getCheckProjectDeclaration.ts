import { CheckProjectDeclaration } from '../../domain/objects/CheckProjectDeclaration';
import { readFileAsync } from '../../utils/fileio/readFileAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';
import { UnexpectedCodePathError } from '../UnexpectedCodePathError';
import { getCheckFileDeclaration } from './getCheckFileDeclaration/getCheckFileDeclaration';

export const getCheckProjectDeclaration = async ({
  declaredProjectDirectory,
}: {
  declaredProjectDirectory: string;
}): Promise<CheckProjectDeclaration> => {
  // grab name from the directory
  const bestPracticeName = (new RegExp(/\/([\w-]+)\/best-practice$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  const badPracticeName = (new RegExp(/bad-practices\/([\w-]+)$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  const name = bestPracticeName ?? badPracticeName;
  if (!name)
    throw new UnexpectedCodePathError(
      `neither best-practice name nor bad-practice name was extractable from the declared project directory '${declaredProjectDirectory}'`,
    );

  // grab paths to _all_ files in this dir (not just at root level)
  const paths = await listFilesInDirectory(declaredProjectDirectory);

  // grab the meta files (i.e., path matches `${projectRoot}/.declapract.*`)
  const metaFilePaths = paths.filter((path) => new RegExp(/^\.declapract\./).test(path));

  // group all of the other files by main file name (i.e., key = filePath.replace('.declapract.ts$', ''))
  const projectFilePaths = [
    ...new Set(
      paths.filter((path) => !metaFilePaths.includes(path)).map((path) => path.replace(/\.declapract\.ts$/, '')),
    ),
  ];

  // for each "main file", get the CheckFileDefinition, now that we have all the files defined for it
  const checks = await Promise.all(
    projectFilePaths.map((declaredFileCorePath) =>
      getCheckFileDeclaration({ declaredProjectDirectory, declaredFileCorePath }),
    ),
  );

  // get readme contents, if readme defined
  const readme = metaFilePaths.includes('.declapract.readme.md')
    ? await readFileAsync({ filePath: `${declaredProjectDirectory}/.declapract.readme.md` })
    : null;

  // return the project def
  return new CheckProjectDeclaration({
    name,
    dir: declaredProjectDirectory,
    checks,
    readme,
  });
};
