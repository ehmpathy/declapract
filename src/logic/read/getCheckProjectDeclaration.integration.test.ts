import { getCheckProjectDeclaration } from './getCheckProjectDeclaration';

describe('getCheckProjectDeclarationFromDirectory', () => {
  it('should define correctly when only one root level file check and a readme', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${__dirname}/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
  it('should define correctly when many files in many levels', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${__dirname}/../__test_assets__/exampleDeclarations/practices/terraform/best-practice`,
    });
    expect(project.checks.length).toEqual(7);
    expect(project).toMatchSnapshot();
  });
  it.only('should define correctly for a bad practice', async () => {
    const project = await getCheckProjectDeclaration({
      declaredProjectDirectory: `${__dirname}/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
});
