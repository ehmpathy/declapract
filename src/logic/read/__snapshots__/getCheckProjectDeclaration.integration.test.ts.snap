// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCheckProjectDeclarationFromDirectory should define correctly for a bad practice 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment/package.json",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/bad-practices/moment",
  "name": "moment",
  "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
  "test": null,
}
`;

exports[`getCheckProjectDeclarationFromDirectory should define correctly when many files in many levels 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/.gitignore",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/.terraform-version",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/dev/main.tf",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/dev/versions.tf",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/prod/main.tf",
      "required": true,
      "type": "CONTAINS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/environments/prod/versions.tf",
      "required": true,
      "type": "EQUALS",
    },
    CheckFileDeclaration {
      "check": [Function],
      "fix": [Function],
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/provision/terraform/product/variables.tf",
      "required": true,
      "type": "EQUALS",
    },
  ],
  "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice",
  "name": "terraform",
  "readme": "best pattern found for terraform so far is to put different environments into their own directory - and just have them reference a shared product dir
",
  "test": TestSuiteReference {
    "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/terraform/best-practice/.declapract.test.ts",
  },
}
`;

exports[`getCheckProjectDeclarationFromDirectory should define correctly when only one root level file check and a readme 1`] = `
CheckProjectDeclaration {
  "checks": Array [
    CheckFileDeclaration {
      "check": [Function],
      "fix": null,
      "path": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice/package.json",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "dir": "/home/uladkasach/git/more/declapract/src/logic/read/../__test_assets__/exampleDeclarations/practices/dates-and-times/best-practice",
  "name": "dates-and-times",
  "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
  "test": null,
}
`;
